### PR TITLE
dev fibers2

### PR DESCRIPTION
**Pull request by myannyax**

### CODE REVIEW

File: src/coroutines/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,37 @@
+list(APPEND YACLIB_INCLUDES
+  ${YACLIB_INCLUDE_DIR}/coroutines/standalone_coroutine.hpp
+  ${YACLIB_INCLUDE_DIR}/coroutines/context/stack.hpp
+  ${YACLIB_INCLUDE_DIR}/coroutines/context/stack_allocator.hpp
+  ${YACLIB_INCLUDE_DIR}/coroutines/context/stack_view.hpp
+  ${YACLIB_INCLUDE_DIR}/coroutines/context/default_allocator.hpp
+  )
+
+list(APPEND YACLIB_HEADERS
+  # todo: pick based on a system
```
Comment (MBkkt) [inline src/coroutines/CMakeLists.txt:10]: We write todo like this:
```suggestion
  # TODO(myannyax) Pick based on a system
```

File: src/coroutines/context/asm/switch_context_x86_64.S [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    movq        (YACLIB_RBX_INDEX * 8)(%rsi), %rbx
+    movq        (YACLIB_RBP_INDEX * 8)(%rsi), %rbp
+    movq        (YACLIB_R12_INDEX * 8)(%rsi), %r12
+    movq        (YACLIB_R13_INDEX * 8)(%rsi), %r13
+    movq        (YACLIB_R14_INDEX * 8)(%rsi), %r14
+    movq        (YACLIB_R15_INDEX * 8)(%rsi), %r15
+
+    movq        (YACLIB_RSP_INDEX * 8)(%rsi), %rsp
+    movq        (YACLIB_RIP_INDEX * 8)(%rsi), %rcx
+    // for trampoline
```
Comment (MBkkt) [inline src/coroutines/context/asm/switch_context_x86_64.S:32]: https://github.com/YACLib/YACLib/pull/104
Check this

File: src/coroutines/context/execution_context.hpp [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+#include "machine_context.hpp"
+#include <yaclib/coroutines/context/stack_view.hpp>
+
+#include <algorithm>
+
+namespace yaclib {
+
+class ExecutionContext {
```
Comment (MBkkt) [inline src/coroutines/context/execution_context.hpp:9]: Why do you use so many same contexts?

File: src/coroutines/context/setup_stack/setup_stack_x64.hpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include <yaclib/coroutines/context/stack_view.hpp>
+
+namespace yaclib {
+
+using Trampoline = void (*)(void* arg);
+
+struct YaclibFiberMachineContext {
```
Comment (MBkkt) [inline src/coroutines/context/setup_stack/setup_stack_x64.hpp:9]: You should use 
`void* registers[YACLIB_CONTEXT_SIZE];`
And use `switch_context_x86_64.hpp` for addressing

File: src/coroutines/coroutine_impl.cpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
+#include "coroutine_impl.hpp"
+
+namespace yaclib {
+
+void CoroutineImpl::operator()() {
+  Resume();
+}
+void CoroutineImpl::Resume() {
```
Comment (MBkkt) [inline src/coroutines/coroutine_impl.cpp:8]: everywhere
```suggestion
}

void CoroutineImpl::Resume() {
```

**Review by MBkkt — CHANGES_REQUESTED** [review]
Correct comments.
Write at least some benchmark (or better copy it from other libraries).
Remove the boilerplate code and classes.

File: test/unit/coroutines/basic.cpp [hunk 6]
```diff
@@ -0,0 +1,105 @@
+#include "yaclib/fibers/coroutine.hpp"
```
Comment (MBkkt) [inline test/unit/coroutines/basic.cpp:1]: Use `#include <...>`

File: .github/workflows/fibers.yml [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+            cmake_args: '-DYACLIB_DEFINITIONS=/DYACLIB_CI_SLOWDOWN=10',
+          }
+          - {
+            os: macos-latest,
+            cc: clang, cxx: clang++,
+            cxx_standard: 17,
+            cmake_args: '-DYACLIB_DEFINITIONS=/DYACLIB_CI_SLOWDOWN=10',
+          }
+
+          - {
```
Comment (MBkkt) [inline .github/workflows/fibers.yml:160]: Comment out windows and write todo

File: include/yaclib/fibers/coroutine.hpp [hunk 8]
```diff
@@ -0,0 +1,31 @@
+#pragma once
+
+#include "default_allocator.hpp"
```
Comment (MBkkt) [inline include/yaclib/fibers/coroutine.hpp:3]: Plz use `include <...>`

File: include/yaclib/fibers/coroutine_base.hpp [hunk 9]
```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include "stack_view.hpp"
```
Comment (MBkkt) [inline include/yaclib/fibers/coroutine_base.hpp:3]: include <>

File: include/yaclib/fibers/coroutine_base.hpp [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+using Routine = yaclib::util::IFuncPtr;
+
+/**
+ * Base coroutine class
+ */
+class CoroutineBase {
+ public:
+  CoroutineBase(const StackView& stack_view, Routine routine) : _routine(std::move(routine)) {
+    _context.Setup(stack_view, Trampoline, this);
```
Comment (MBkkt) [inline include/yaclib/fibers/coroutine_base.hpp:20]: Should be in cpp

File: include/yaclib/fibers/detail/execution_context.hpp [hunk 11]
```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+#include "yaclib/fibers/stack_view.hpp"
```
Comment (MBkkt) [inline include/yaclib/fibers/detail/execution_context.hpp:3]: include <>

File: src/fibers/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  ${YACLIB_INCLUDE_DIR}/fibers/stack.hpp
+  ${YACLIB_INCLUDE_DIR}/fibers/stack_allocator.hpp
+  ${YACLIB_INCLUDE_DIR}/fibers/stack_view.hpp
+  ${YACLIB_INCLUDE_DIR}/fibers/default_allocator.hpp
+  ${YACLIB_INCLUDE_DIR}/fibers/detail/execution_context.hpp
+  )
+
+list(APPEND YACLIB_HEADERS
+  # TODO(myannyax): pick based on a system
+  ${CMAKE_CURRENT_SOURCE_DIR}/context/asm/switch_context_x86_64.hpp
```
Comment (MBkkt) [inline src/fibers/CMakeLists.txt:13]: move to }/fiber/asm...

File: src/fibers/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,32 @@
+list(APPEND YACLIB_INCLUDES
+  ${YACLIB_INCLUDE_DIR}/fibers/coroutine.hpp
+  ${YACLIB_INCLUDE_DIR}/fibers/coroutine_base.hpp
+  ${YACLIB_INCLUDE_DIR}/fibers/stack.hpp
+  ${YACLIB_INCLUDE_DIR}/fibers/stack_allocator.hpp
+  ${YACLIB_INCLUDE_DIR}/fibers/stack_view.hpp
+  ${YACLIB_INCLUDE_DIR}/fibers/default_allocator.hpp
+  ${YACLIB_INCLUDE_DIR}/fibers/detail/execution_context.hpp
```
Comment (MBkkt) [inline src/fibers/CMakeLists.txt:8]: fiber instead of fibers

File: src/fibers/context/default_allocator.cpp [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    }
+    _stack_size_pages = pages + 1;
+  }
+}
+
+size_t DefaultAllocator::GetMinStackSize() {
+  return (_stack_size_pages - 1) * kPageSize;
+}
+
+DefaultAllocator default_allocator_instance;
```
Comment (MBkkt) [inline src/fibers/context/default_allocator.cpp:65]: Should be named gDefaultAllocator

File: src/fibers/context/execution_context.cpp [hunk 15]
```diff
@@ -0,0 +1,24 @@
+#include "yaclib/fibers/detail/execution_context.hpp"
+// TODO(myannyax) if
+#include "setup_stack_x64.hpp"
+
+extern "C" void yaclib_switch_context(void* from_context, void* to_context);
+
+namespace yaclib {
+
+ExecutionContext::ExecutionContext() : _context(new void*[kAsmContextSize]) {
```
Comment (MBkkt) [inline src/fibers/context/execution_context.cpp:9]: - [ ] Remove allocation.

Btw wrong size, you alloc 8*size, where size is count of registers * 8

Use char[kAsmContextSize]; and cast it to void**

File: src/fibers/context/setup_stack_x64.cpp [hunk 16]
```diff
@@ -0,0 +1,23 @@
+#include "setup_stack_x64.hpp"
```
Comment (MBkkt) [inline src/fibers/context/setup_stack_x64.cpp:1]: include <>

File: src/fibers/context/setup_stack_x64.hpp [hunk 17]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include "yaclib/fibers/detail/execution_context.hpp"
+#include "yaclib/fibers/stack_view.hpp"
+
+namespace yaclib {
+
+// TODO(myannyax) check how much saving on stack is better
+extern const size_t kAsmContextSize;
```
Comment (MBkkt) [inline src/fibers/context/setup_stack_x64.hpp:9]: Use inline constexpr

File: src/fibers/context/setup_stack_x64.cpp [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+
+#include "asm/switch_context_x86_64.hpp"
+
+namespace yaclib {
+
+extern "C" void yaclib_trampoline();
```
Comment (MBkkt) [inline src/fibers/context/setup_stack_x64.cpp:7]: Remove from namespace

File: src/fibers/context/setup_stack_x64.cpp [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+
+static const int kAlignment = 16;
```
Comment (MBkkt) [inline src/fibers/context/setup_stack_x64.cpp:9]: constexpr

File: src/fibers/coroutine.cpp [hunk 20]
```diff
@@ -0,0 +1,34 @@
+#include "yaclib/fibers/coroutine.hpp"
```
Comment (MBkkt) [inline src/fibers/coroutine.cpp:1]: <>

File: test/CMakeLists.txt [hunk 21]
```diff
@@ -56,6 +56,7 @@ set(YACLIB_UNITS
   unit/executor/thread_factory
   unit/executor/thread_pool
   unit/executor/serial
+  unit/coroutines/basic
```
Comment (MBkkt) [inline test/CMakeLists.txt:59]: unit/fiber/basic

**Review by MBkkt — CHANGES_REQUESTED** [review]

File: src/fibers/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+list(APPEND YACLIB_HEADERS
+  # TODO(myannyax): pick based on a system
+  ${CMAKE_CURRENT_SOURCE_DIR}/context/asm/switch_context_x86_64.hpp
+  ${CMAKE_CURRENT_SOURCE_DIR}/context/setup_stack_x64.hpp
+  )
+
+SET(CMAKE_ASM_FLAGS "${CFLAGS} -x assembler-with-cpp")
+
+list(APPEND YACLIB_SOURCES
+  ${CMAKE_CURRENT_SOURCE_DIR}/coroutine_base.cpp
```
[the remark this answers is not in the document]
Comment (myannyax) [reply src/fibers/CMakeLists.txt:20]: CMAKE_CURRENT_SOURCE_DIR already contain fiber 🤔

[the remark this answers is not in the document]
Comment (MBkkt) [reply src/fibers/CMakeLists.txt:20]: Yeah

**Review by MBkkt — APPROVED** [review]

**Review by MBkkt — APPROVED** [review]

**Pull request merged**